sw/airborne/modules/relative_localization_filter/relative_localization_filter.c

```diff
@@ -0,0 +1,111 @@
+ * any later version.
+ *
+ * paparazzi is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with paparazzi; see the file COPYING.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+/**
+ * @file "modules/relativelocalizationfilter/relativelocalizationfilter.c"
+ * @author Mario Coppola
+ * Relative Localization Filter for collision avoidance between drones
+ */
+
+#include "relative_localization_filter.h"
+#include "subsystems/datalink/telemetry.h"
+#include "state.h" // To get current states
+
+#include "modules/datalink/extra_pprz_dl.h"
+#include "subsystems/abi.h"
+
+#include "math/pprz_algebra.h"
+#include "math/pprz_algebra_int.h"
+
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "discrete_ekf.h"
+
+#ifndef RL_NUAVS
+#define RL_NUAVS 4 // Maximum expected number of other UAVs
+#endif
+
+int32_t id_array[RL_NUAVS]; // array of UWB IDs of all drones
+uint32_t latest_update_time[RL_NUAVS];
+uint8_t number_filters; // the number of filters running in parallel
+struct discrete_ekf ekf_rl[RL_NUAVS]; 
+float range_array[RL_NUAVS]; // an array to store the ranges at which the other MAVs are
+uint8_t pprzmsg_cnt; // a counter to send paparazzi messages, which are sent in rotation
+
+static abi_event range_communication_event;
+static void range_msg_callback(uint8_t sender_id __attribute__((unused)),
+                               uint8_t ac_id, float range, float tracked_v_north, float tracked_v_east, float tracked_h)
+{
+  int idx = -1; // Initialize the index of all tracked drones (-1 for null assumption of no drone found)
+
+  // Check if a new aircraft ID is present, if it's a new ID we start a new EKF for it.
+  if (!int32_vect_find(id_array, ac_id, &idx, RL_NUAVS) &&
+      (number_filters < RL_NUAVS)) {
+    id_array[number_filters] = ac_id;
+    discrete_ekf_new(&ekf_rl[number_filters]);
+    number_filters++;
+  } else if (idx != -1) {
+    range_array[idx] = range;
+    ekf_rl[idx].dt = (get_sys_time_usec() - latest_update_time[idx]) / pow(10, 6); // Update the time between messages
+
+    // Measurement Vector Z = [range owvVx(NED) ownVy(NED) tracked_v_north(NED) tracked_v_east(NED) dh]
+    float Z[EKF_M] = {range, stateGetSpeedEnu_f()->y, stateGetSpeedEnu_f()->x, tracked_v_north, tracked_v_east, tracked_h - stateGetPositionEnu_f()->z};
+
+    discrete_ekf_predict(&ekf_rl[idx]);
+    discrete_ekf_update(&ekf_rl[idx], Z);
+  }
+
+  latest_update_time[idx] = get_sys_time_usec();
+};
+
+// It is best to send the data of each tracked drone separately to avoid overloading
+static void send_relative_localization_data(struct transport_tx *trans, struct link_device *dev)
+{
+  pprzmsg_cnt++;
+  if (pprzmsg_cnt >= number_filters) {
+    pprzmsg_cnt = 0;
+  }
```

Comment:
sorry to nag here but what if number_filters=0? Won't this message will contain gibberish.